A38-admin-interface-api.md

```diff
@@ -0,0 +1,200 @@
+
+To help communicate the scenarios, we can define all services into several types, from the view of administrative functions:
+
+* **Primary service**: the user-provided services that handle the application logic;
+* **Binary-global service**: the services that expose the state of the entire gRPC library, and they serve the same information regardless of which server they have been registered (e.g., Channelz, CSDS);
+* **Per-server service**: the services that distribute the state of other registered services on the same server (e.g., Health, Reflection).
+
+The binary-global services contain sensitive information about traffic, configuration, even security, hence we hope users can set up the services securely in our recommended usage or the upcoming new API.
+
+The recommended way of securing the admin service port is binding it to either `localhost` or using a `UDS` port.
+
+
+## Proposal
+
+There are several different options for the convenience API design to help create a more usable admin interface API. Each option is a tradeoff between simplicity to use and flexibility. The simplicity to use is defined by lines of code to create a serving gRPC server with admin services, and the additional cognitive cost to learn the new API or manage dependencies.
+
+The flexibility is defined by how can users provide additional server creation options. To date, users can specify one or more **listening addresses, credentials, TCP/HTTP config, codec, interceptors and service implementations** during the server creation process. Not to mention the potential  impact on the threading model during the start of the server. If we want to bundle as many options as possible into one line of code, we will inevitably sacrifice some flexibility.
+
+Options to create an admin server are ordered from the most flexible to the least:
+
+1. The control group: using existing API to create the server;
+2. A function that combines admin services from different packages; **<-RECOMMENDED BY THIS PROPOSAL**
+3. A method of the gRPC server class that adds admin services;
+4. A function that creates the gRPC server object bound with admin services;
+5. A function that creates and starts the gRPC server with admin services.
+
+This proposal will recommend option 2 as the admin interface API (for other alternatives, see rationale section below). The usage snippet of admin interface API in each language:
+
+
+### Java
+
+#### Create An Admin Server With Existing gRPC Java API:
+
+```java
+server = ServerBuilder.forPort(50051)
+        .useTransportSecurity(certChainFile, privateKeyFile)
+        .addService(new ChannelzService())
+        .addService(new CsdsService())
+        .build()
+        .start();
+server.awaitTermination();
+```
+
+
+#### Create An Admin Server With gRPC Java Admin Interface API:
+
+```java
+server = ServerBuilder.forPort(50051)
+        .useTransportSecurity(certChainFile, privateKeyFile)
+        .addServices(AdminInterface.getStandardServices())
```

Comment:
We'll also need `addServices()` to be added to the builder

Reply:
Yes. This method needs to be added.